c++/src/ColumnWriter.cc

```diff
@@ -940,24 +940,98 @@ namespace orc {
     lengthEncoder->recordPosition(rowIndexPosition.get());
   }
 
+  struct Utf8Utils {
+    /**
+     * Counts how many utf-8 chars of the inout data
+     */
+    static uint64_t charLength(const char * data, uint64_t offset, uint64_t length) {
+      uint64_t chars = 0;
+      for (uint64_t i = 0; i < length; i++) {
+        if (isUtfStartByte(data[offset + i])) {
+          chars++;
+        }
+      }
+      return chars;
+    }
+
+    /**
+     * Return the number of bytes required to read at most
+     * maxLength characters in full from a utf-8 encoded byte array provided
+     * by data[offset:offset+length]. This does not validate utf-8 data, but
+     * operates correctly on already valid utf-8 data.
+     *
+     * @param maxCharLength number of bytes required
+     * @param data the bytes of UTF-8
+     * @param offset the first byte location
+     * @param length the length of data to truncate
+     */
+    static uint64_t truncateBytesTo(uint64_t maxCharLength,
+                                    const char * data,
+                                    uint64_t offset,
+                                    uint64_t length) {
+      uint64_t chars = 0;
+      if (length <= maxCharLength) {
```

Comment:
the length param means the length after offset, so the if statement is correct.
I just copied this from java side, yes I can remove offset param to make it cleaner because C++ side is much simpler here.